modules/xfem/src/base/XFEMGeometricCut3D.C

```diff
@@ -0,0 +1,168 @@
+    {
+      std::vector<unsigned int> cut_edges;
+      std::vector<Real> cut_pos;
+
+      for (unsigned int j=0; j < n_face_nodes; j++)
+      {
+        unsigned int jplus1(j < (n_face_nodes-1) ? (j+1) : 0);
+        unsigned int node_id1 = 0;
+        unsigned int node_id2 = 0;
+        if (elem->n_nodes() == 8)  // hex
+        {
+          node_id1 = hex_indices[i][j];
+          node_id2 = hex_indices[i][jplus1];
+        }else if (elem->n_nodes() == 4) // tet
+        {
+          node_id1 = tet_indices[i][j];
+          node_id2 = tet_indices[i][jplus1];
+        }
+        Node *node1 = elem->get_node(node_id1);
+        Node *node2 = elem->get_node(node_id2);
+        Point p1((*node1)(0), (*node1)(1), (*node1)(2));
+        Point p2((*node2)(0), (*node2)(1), (*node2)(2));
+
+        Point pint(0.0,0.0,0.0);
+        if (intersectWithEdge(p1,p2,pint))
+        {
+          cut_edges.push_back(j);
+          cut_pos.push_back(getRelativePosition(p1,p2,pint));
+        }
+      }
+
+      if (cut_edges.size() == 2)
+      {
+       cut_elem = true;
+       cutFace mycut;
+       mycut.face_id = i;
+       mycut.face_edge.push_back(cut_edges[0]);
+       mycut.face_edge.push_back(cut_edges[1]);
+       mycut.position.push_back(cut_pos[0]);
+       mycut.position.push_back(cut_pos[1]);
+       cutFaces.push_back(mycut);
+     }
+
+   }
+
+   return cut_elem;
+}
+
+bool
+XFEMGeometricCut3D::cutFragmentByGeometry(std::vector<std::vector<Point> > & /*frag_edges*/,
+                                          std::vector<cutEdge> & /*cutEdges*/,
+                                          Real /*time*/)
+{
+  mooseError("invalid method for 3D mesh cutting");
+  return false;
+}
+
+
+bool
+XFEMGeometricCut3D::cutFragmentByGeometry(std::vector<std::vector<Point> > & /*frag_faces*/,
+                                          std::vector<cutFace> & /*cutFaces*/,
+                                          Real /*time*/)
+{
+  //TODO: Need this for branching in 3D
+  mooseError("cutFragmentByGeometry not yet implemented for 3D mesh cutting");
+  return false;
+}
+
+bool
+XFEMGeometricCut3D::intersectWithEdge(Point p1, Point p2, Point &pint)
+{
+  bool has_intersection = false;
+  double plane_point[3] = {_center(0), _center(1), _center(2)};
+  double plane_normal[3] = {_normal(0), _normal(1), _normal(2)};
+  double edge_point1[3] = {p1(0), p1(1), p1(2)};
+  double edge_point2[3] = {p2(0), p2(1), p2(2)};
+  double cut_point[3] = {0.0,0.0,0.0};
+
+  if (plane_normal_line_exp_int_3d(plane_point, plane_normal, edge_point1, edge_point2, cut_point) == 1) {
+    Point temp_p(cut_point[0], cut_point[1], cut_point[2]);
+    if ( isInsideCutPlane(temp_p) && isInsideEdge(p1, p2, temp_p) )
+    {
+      pint = temp_p;
+      has_intersection = true;
+    }
+  }
+  return has_intersection;
+}
+
+bool
+XFEMGeometricCut3D::isInsideEdge(Point p1, Point p2, Point p)
+{
+  Point p1_to_p2 = p2 - p1;
+  Point p_to_p1 = p1 - p;
+  Point p_to_p2 = p2 - p;
```

Comment:
these temporary expression names are longer than the terms they represent. Just ditch them.